File: tools/build_dashboard.mjs
// Pre-transpile the dashboard JSX -> plain JS so the browser loads compiled
// React.createElement code directly (no in-browser @babel/standalone). This
// removes the ~10s transpile-on-load (the blank-page window) and the unpkg
// babel download.
//
// Usage:  node tools/build_dashboard.mjs
// Requires @babel/standalone (npm i --no-save @babel/standalone). The wrapper
// tools/build_dashboard.py installs it on demand and runs this.
//
// Re-run after editing any gui/dashboard/*.jsx file. The generated *.js are
// committed so the dashboard works without a build step at serve time.
import { readFileSync, writeFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import * as babel from '@babel/standalone';

const here = dirname(fileURLToPath(import.meta.url));
const dashDir = join(here, '..', 'gui', 'dashboard');

// Order matters at load time; transpile each independently (classic runtime,
// React/ReactDOM are globals from the UMD scripts).
const FILES = ['shared.jsx', 'direction_b.jsx', 'direction_b_pages.jsx', 'app.jsx'];

let ok = 0;
for (const f of FILES) {
  const src = readFileSync(join(dashDir, f), 'utf8');
  const out = babel.transform(src, {
    presets: ['react'],
    filename: f,
    compact: false,
    retainLines: false,
  }).code;
  const dest = f.replace(/\.jsx$/, '.js');
  const banner = `// GENERATED from ${f} by tools/build_dashboard.mjs — do not edit; edit the .jsx and rebuild.\n`;
  writeFileSync(join(dashDir, dest), banner + out, 'utf8');
  console.log(`  ${f} -> ${dest} (${out.length} bytes)`);
  ok++;
}
console.log(`build_dashboard: transpiled ${ok}/${FILES.length} files`);

File: tools/build_dashboard.py
#!/usr/bin/env python3
"""Build the dashboard: transpile gui/dashboard/*.jsx -> *.js (no in-browser Babel).

Ensures @babel/standalone is available (npm i --no-save) then runs the node
transpiler tools/build_dashboard.mjs. Run this after editing any dashboard .jsx;
the generated .js are committed so the server needs no build step at runtime.

    python3 tools/build_dashboard.py
"""
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent


def main() -> int:
    # Install the transpiler (no-save: not a project dep, just a build tool).
    print("ensuring @babel/standalone is available…")
    r = subprocess.run(["npm", "install", "--no-save", "@babel/standalone"],
                       cwd=ROOT, capture_output=True, text=True, shell=(sys.platform == "win32"))
    if r.returncode != 0:
        print("npm install failed:\n" + (r.stderr or r.stdout)[-1000:])
        return 1
    print("transpiling JSX…")
    r = subprocess.run(["node", str(ROOT / "tools" / "build_dashboard.mjs")],
                       cwd=ROOT, capture_output=True, text=True, shell=(sys.platform == "win32"))
    sys.stdout.write(r.stdout)
    if r.returncode != 0:
        sys.stderr.write(r.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
